src/make_watertight/Gen.cpp

```diff
@@ -0,0 +1,2117 @@
+    moab::Range triangles;
+    rval = MBI()->get_entities_by_dimension(surfaces[i], 2, triangles);
+    if (moab::MB_SUCCESS != rval) {
+      return rval;
+    }
+    if (!triangles.all_of_type(moab::MBTRI)) {
+      std::cout << "WARNING: Surface " << geom_id_by_handle(surfaces[i])
+                << " contains non-triangle elements. Volume calculation may be incorrect."
+                << std::endl;
+      triangles.clear();
+      rval = MBI()->get_entities_by_type(surfaces[i], moab::MBTRI, triangles);
+      if (moab::MB_SUCCESS != rval) {
+        return rval;
+      }
+    }
+
+    // calculate signed volume beneath surface (x 6.0)
+    double surf_sum = 0.0;
+    const moab::EntityHandle* conn;
+    int len;
+    moab::CartVect coords[3];
+    for (moab::Range::iterator j = triangles.begin(); j != triangles.end(); ++j) {
+      rval = MBI()->get_connectivity(*j, conn, len, true);
+      if (moab::MB_SUCCESS != rval)
+        return rval;
+      assert(3 == len);
+      rval = MBI()->get_coords(conn, 3, coords[0].array());
+      if (moab::MB_SUCCESS != rval)
+        return rval;
+
+      coords[1] -= coords[0];
+      coords[2] -= coords[0];
+      surf_sum += (coords[0] % (coords[1] * coords[2]));
+    }
+    result += senses[i] * surf_sum;
+  }
+
+  result /= 6.0;
+  if (debug)
+    std::cout << result << std::endl;
+  return moab::MB_SUCCESS;
+}
+
+/// Calculate the signed volumes beneath the surface (x 6.0). Use the triangle's
+///   cannonical sense. Do not take sense tags into account. Code taken from
+///   DagMC::measure_volume.
+moab::ErrorCode Gen::get_signed_volume(const moab::EntityHandle surf_set, double& signed_volume) {
+  moab::ErrorCode rval;
+  moab::Range tris;
+  rval = MBI()->get_entities_by_type(surf_set, moab::MBTRI, tris);
+  if (moab::MB_SUCCESS != rval)
+    return rval;
+  signed_volume = 0.0;
+  const moab::EntityHandle* conn;
+  int len;
+  moab::CartVect coords[3];
+  for (moab::Range::iterator j = tris.begin(); j != tris.end(); ++j) {
+    rval = MBI()->get_connectivity(*j, conn, len, true);
+    if (moab::MB_SUCCESS != rval)
+      return rval;
+    assert(3 == len);
+    rval = MBI()->get_coords(conn, 3, coords[0].array());
+    if (moab::MB_SUCCESS != rval)
+      return rval;
+
+    coords[1] -= coords[0];
+    coords[2] -= coords[0];
+    signed_volume += (coords[0] % (coords[1] * coords[2]));
+  }
+  return moab::MB_SUCCESS;
+}
+
+moab::ErrorCode Gen::measure(const moab::EntityHandle set, const moab::Tag geom_tag, double& size, bool debug,  bool verbose) {
+  moab::ErrorCode result;
+  int dim;
+  result = MBI()->tag_get_data(geom_tag, &set, 1, &dim);
+  assert(moab::MB_SUCCESS == result);
+  if (0 == dim) {
+    std::cout << "measure: cannot measure vertex" << std::endl;
+    return moab::MB_FAILURE;
+
+  } else if (1 == dim) {
+    std::vector<moab::EntityHandle> vctr;
+    result = get_meshset(set, vctr);
+    assert(moab::MB_SUCCESS == result);
+    size = length(vctr);
```

Comment:
Bug: `std::vector` length is checked by calling `.size()`.

Reply:
The return value here is the output of a measurement function in `Gen.hpp` which computes the length of a curve, not the size of the vector. I'll update the variable name here to make that more clear.